concordium-consensus/src/Concordium/GlobalState/Instance.hs

```diff
@@ -93,64 +112,61 @@ data Instance = InstanceV0 (InstanceV GSWasm.V0)
 
--- |Helper function for JSON encoding an 'Instance'.
-instancePairs :: KeyValue kv => Instance -> [kv]
-{-# INLINE instancePairs #-}
-instancePairs (InstanceV0 InstanceV{..}) =
-    [ "model" .= _instanceVModel,
-      "owner" .= instanceOwner  _instanceVParameters,
-      "amount" .= _instanceVAmount,
-      "methods" .= instanceReceiveFuns  _instanceVParameters,
-      "name" .= instanceInitName  _instanceVParameters,
-      "sourceModule" .= GSWasm.miModuleRef (instanceModuleInterface  _instanceVParameters),
-      "version" .= Wasm.V0
-    ]
-instancePairs (InstanceV1 InstanceV{..}) =
-    [ "owner" .= instanceOwner  _instanceVParameters,
-      "amount" .= _instanceVAmount,
-      "methods" .= instanceReceiveFuns  _instanceVParameters,
-      "name" .= instanceInitName  _instanceVParameters,
-      "sourceModule" .= GSWasm.miModuleRef (instanceModuleInterface  _instanceVParameters),
-      "version" .= Wasm.V1
-    ]
-    
-
--- |JSON instance to support consensus queries.
-instance ToJSON Instance where
-    toJSON inst = object $ instancePairs inst
-    toEncoding inst = pairs $ mconcat $ instancePairs inst
-
+-- |Compute the hash of the instance parameters.
 makeInstanceParameterHash :: ContractAddress -> AccountAddress -> ModuleRef -> Wasm.InitName -> H.Hash
 makeInstanceParameterHash ca aa modRef conName = H.hashLazy $ runPutLazy $ do
         put ca
         put aa
         put modRef
         put conName
 
-makeInstanceHash' :: H.Hash -> Wasm.ContractState -> Amount -> H.Hash
-makeInstanceHash' paramHash conState a = H.hashLazy $ runPutLazy $ do
+-- |Construct the hash of a basic instance from the __hash of the parameters__, the state, and amount for a V0 instance.
+makeInstanceHashV0' :: H.Hash -> InstanceStateV GSWasm.V0 -> Amount -> H.Hash
+makeInstanceHashV0' paramHash (InstanceStateV0 conState) a = H.hashLazy $ runPutLazy $ do
         put paramHash
         putByteString (H.hashToByteString (getHash conState))
         put a
 
-makeInstanceHash :: InstanceParameters v -> Wasm.ContractState -> Amount -> H.Hash
-makeInstanceHash params = makeInstanceHash' (instanceParameterHash params)
+-- |Construct the hash of a basic instance from the instance parameters, the state, and amount for a V0 instance.
+makeInstanceHashV0 :: InstanceParameters v -> InstanceStateV GSWasm.V0 -> Amount -> H.Hash
+makeInstanceHashV0 = makeInstanceHashV0' . instanceParameterHash
+
+-- |Construct the hash of a basic instance from the __hash of the parameters__, the state, and amount for a V1 instance.
+makeInstanceHashV1' :: H.Hash -> InstanceStateV GSWasm.V1 -> Amount -> H.Hash
+makeInstanceHashV1' paramHash (InstanceStateV1 conState) a = H.hashLazy $ runPutLazy $ do
+        -- We put an explicit version for V1. This is not present in V0 for legacy reasons, and we cannot change it now,
+        -- but it would be best if it was present.
+        put Wasm.V1
+        put paramHash
+        put (getHash conState :: SHA256.Hash)
+        put a
```

Comment:
The hashing we have in many places is not entirely consistent. Sometimes there are hash collisions for different objects, so the context is needed to disambiguate. 

I removed the version now since it is actually already part of the moduleref, which is part of the parameter hash.